File: driver/dev/src/pie_driver_dev/shmem_schema.py
"""Zero-copy flat schema for the shmem fast path.

Iteration 2 implemented `parse_request` (Rust → Python).
Iteration 3 adds `write_response` (Python → Rust) — the common-case
token-only response is laid out as a tiny flat header + concatenated
tokens, eliminating the `create_responses` Python loop and msgpack
serialization on the worker side.

Response wire format:

    [Header: 16 bytes]
      0:  u32 magic    = 0x42504953 ('BPIS')
      4:  u32 mode     0 = flat token-only schema
                       1 = msgpack-fallback (advanced path)
      8:  u32 num_requests
      12: u32 total_tokens

    [Mode 0 body]
      u32[num_requests] : per-request token count
      u32[total_tokens] : concatenated tokens

    [Mode 1 body]
      msgpack-encoded {"results": [...]}

Request layout matches `runtime/src/shmem_schema.rs`. We build numpy views
into the shmem buffer (no copy) and assemble a kwargs dict that
`_handle_fire_batch` already expects (msgpack-decoded bytes are reinterpreted
via np.frombuffer in `Batch.__init__`; we provide compatible bytes-typed
values).
"""

from __future__ import annotations

import numpy as np

MAGIC = 0x42504951  # 'BPIQ'
# v2 added A_PREDICT_FLAGS (u8 per request) for pass-level speculative
# execution. See SPECULATIVE_EXECUTION_DESIGN.md.
SCHEMA_VERSION = 2

# Response schema constants
RESP_MAGIC = 0x42504953  # 'BPIS'
RESP_HEADER_SIZE = 16
RESP_MODE_FLAT = 0
RESP_MODE_MSGPACK = 1
HEADER_SIZE = 512
NUM_ARRAYS = 29
FIXED_HEADER = 32

# Array indices (must match Rust).
A_TOKEN_IDS = 0
A_POSITION_IDS = 1
A_KV_PAGE_INDICES = 2
A_KV_PAGE_INDPTR = 3
A_KV_LAST_PAGE_LENS = 4
A_QO_INDPTR = 5
A_FLATTENED_MASKS = 6
A_MASK_INDPTR = 7
A_LOGIT_MASKS = 8
A_LOGIT_MASK_INDPTR = 9
A_SAMPLING_INDICES = 10
A_SAMPLING_INDPTR = 11
A_SAMPLER_TEMPERATURES = 12
A_SAMPLER_TOP_K = 13
A_SAMPLER_TOP_P = 14
A_SAMPLER_MIN_P = 15
A_SAMPLER_TYPES = 16
A_SAMPLER_SEEDS = 17
A_REQUEST_NUM_SAMPLERS = 18
A_SAMPLER_LABEL_IDS = 19
A_SAMPLER_LABEL_INDPTR = 20
A_ADAPTER_INDICES = 21  # i64
A_ADAPTER_SEEDS = 22  # i64
A_SPEC_TOKEN_IDS = 23
A_SPEC_POSITION_IDS = 24
A_SPEC_INDPTR = 25
A_OUTPUT_SPEC_FLAGS = 26  # u8
A_CONTEXT_IDS = 27  # u64
A_PREDICT_FLAGS = 28  # u8 (v2)

ELEM_SIZE = [
    4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
    4, 4, 4, 4, 4, 4, 4, 4, 4, 8, 8, 4, 4, 4, 1, 8, 1,
]


def parse_request(view: memoryview) -> dict:
    """Parse a serialized BatchedForwardPassRequest from the given memoryview.

    Returns a dict whose keys match what `_handle_fire_batch` expects (i.e.
    the same keys msgpack would produce from the existing wire format).

    Byte buffers in the returned dict are *views* into the shmem region —
    Python must not retain them past the time the slot is released, but
    they are consumed by `Batch(args, ...)` which copies what it needs into
    numpy arrays before returning from the handler.
    """
    if len(view) < HEADER_SIZE:
        raise ValueError(f"buffer too small ({len(view)} < {HEADER_SIZE})")

    raw = view if isinstance(view, memoryview) else memoryview(view)
    # numpy view of the header to read u32s/u64s
    header = np.frombuffer(raw, dtype=np.uint8, count=HEADER_SIZE)

    magic = int(header[0:4].view(np.uint32)[0])
    if magic != MAGIC:
        raise ValueError(f"shmem schema magic mismatch: 0x{magic:08x} != 0x{MAGIC:08x}")
    schema_version = int(header[4:8].view(np.uint32)[0])
    if schema_version != SCHEMA_VERSION:
        raise ValueError(f"shmem schema version mismatch: {schema_version} != {SCHEMA_VERSION}")
    device_id = int(header[8:12].view(np.uint32)[0])
    flags = int(header[12:16].view(np.uint32)[0])
    num_arrays = int(header[16:20].view(np.uint32)[0])
    if num_arrays != NUM_ARRAYS:
        raise ValueError(f"shmem num_arrays {num_arrays} != {NUM_ARRAYS}")

    single_token_mode = bool(flags & 1)

    # Read offset/len table.
    table = header[FIXED_HEADER:FIXED_HEADER + NUM_ARRAYS * 8].view(np.uint32)
    offsets = table[0::2].copy()  # small (<256B) — copy is fine
    lens = table[1::2].copy()

    def slice_bytes(idx: int) -> bytes:
        off = int(offsets[idx])
        n = int(lens[idx]) * ELEM_SIZE[idx]
        # Return as bytes (immutable copy) — mirrors what msgpack delivers.
        # The Batch constructor calls np.frombuffer(data, dtype=...) which
        # works with bytes or buffer-protocol types. To keep zero-copy, we
        # return a memoryview slice; Batch's np.frombuffer(memoryview) is
        # zero-copy.
        return raw[off:off + n]

    _I64_MIN = np.iinfo(np.int64).min

    def slice_adapter_seeds(idx: int) -> list:
        off = int(offsets[idx])
        n = int(lens[idx])
        if n == 0:
            return []
        arr = np.frombuffer(raw, dtype=np.int64, count=n, offset=off)
        # Common case: all sentinel ⇒ all None.
        if (arr == _I64_MIN).all():
            return [None] * n
        # Otherwise convert via tolist() (one batched call into numpy C).
        py = arr.tolist()
        return [None if v == _I64_MIN else v for v in py]

    def slice_bool_array(idx: int) -> list:
        off = int(offsets[idx])
        n = int(lens[idx])
        if n == 0:
            return []
        arr = np.frombuffer(raw, dtype=np.uint8, count=n, offset=off)
        # Fast paths for all-False / all-True.
        if not arr.any():
            return [False] * n
        if arr.all():
            return [True] * n
        return arr.astype(bool).tolist()

    def slice_u64_array(idx: int) -> list:
        off = int(offsets[idx])
        n = int(lens[idx])
        if n == 0:
            return []
        arr = np.frombuffer(raw, dtype=np.uint64, count=n, offset=off)
        return arr.tolist()  # batched

    # Adapter indices: -1 sentinel.
    def slice_adapter_indices(idx: int) -> list:
        off = int(offsets[idx])
        n = int(lens[idx])
        if n == 0:
            return []
        arr = np.frombuffer(raw, dtype=np.int64, count=n, offset=off)
        if (arr == -1).all():
            return [None] * n
        py = arr.tolist()
        return [None if v == -1 else v for v in py]

    # Match the kwargs that _handle_fire_batch / Batch expects.
    args = {
        "token_ids": slice_bytes(A_TOKEN_IDS),
        "position_ids": slice_bytes(A_POSITION_IDS),
        "kv_page_indices": slice_bytes(A_KV_PAGE_INDICES),
        "kv_page_indptr": slice_bytes(A_KV_PAGE_INDPTR),
        "kv_last_page_lens": slice_bytes(A_KV_LAST_PAGE_LENS),
        "qo_indptr": slice_bytes(A_QO_INDPTR),
        "flattened_masks": slice_bytes(A_FLATTENED_MASKS),
        "mask_indptr": slice_bytes(A_MASK_INDPTR),
        "logit_masks": slice_bytes(A_LOGIT_MASKS),
        "logit_mask_indptr": slice_bytes(A_LOGIT_MASK_INDPTR),
        "sampling_indices": slice_bytes(A_SAMPLING_INDICES),
        "sampling_indptr": slice_bytes(A_SAMPLING_INDPTR),
        "sampler_temperatures": slice_bytes(A_SAMPLER_TEMPERATURES),
        "sampler_top_k": slice_bytes(A_SAMPLER_TOP_K),
        "sampler_top_p": slice_bytes(A_SAMPLER_TOP_P),
        "sampler_min_p": slice_bytes(A_SAMPLER_MIN_P),
        "sampler_types": slice_bytes(A_SAMPLER_TYPES),
        "sampler_seeds": slice_bytes(A_SAMPLER_SEEDS),
        "request_num_samplers": slice_bytes(A_REQUEST_NUM_SAMPLERS),
        "sampler_label_ids": slice_bytes(A_SAMPLER_LABEL_IDS),
        "sampler_label_indptr": slice_bytes(A_SAMPLER_LABEL_INDPTR),
        "adapter_indices": slice_adapter_indices(A_ADAPTER_INDICES),
        "adapter_seeds": slice_adapter_seeds(A_ADAPTER_SEEDS),
        "spec_token_ids": slice_bytes(A_SPEC_TOKEN_IDS),
        "spec_position_ids": slice_bytes(A_SPEC_POSITION_IDS),
        "spec_indptr": slice_bytes(A_SPEC_INDPTR),
        "output_spec_flags": slice_bool_array(A_OUTPUT_SPEC_FLAGS),
        "context_ids": slice_u64_array(A_CONTEXT_IDS),
        "predict_flags": slice_bool_array(A_PREDICT_FLAGS),  # v2
        "single_token_mode": single_token_mode,
        "device_id": device_id,
    }
    return args


_struct_header_pack = None  # lazy init

# Sampler types that produce non-token outputs (dist / logits / logprobs /
# entropies). Presence of any of these in `batch.sampler_types` forces the
# msgpack-fallback path because the flat token-only schema can't carry
# their payloads.
_SPECIAL_SAMPLERS = frozenset({0, 7, 8, 9, 10})


def write_response_v2(sampling_results, batch, dst_buf, msgpack_module=None) -> int:
    """Direct shmem encode bypassing `Batch.create_responses`.

    Detects the fast path (no special-typed samplers, no spec-verify, no
    spec-out drafts) and writes counts + tokens straight into the shmem
    response buffer. Falls back to building a result dict via
    `create_responses` and msgpack-encoding it for the advanced path.

    Saves the per-request Python `ForwardPassResponse` and dict allocations
    that `create_responses` + the calling site otherwise pay for every
    iteration.
    """
    global _struct_header_pack
    if _struct_header_pack is None:
        import struct as _struct
        _struct_header_pack = _struct.Struct("<IIII").pack_into

    sampler_types = batch.sampler_types  # list[int]
    spec_accepted = sampling_results.get("spec_accepted_tokens")
    spec_tokens_all = sampling_results.get("spec_tokens")

    fast_path = (
        spec_accepted is None
        and (spec_tokens_all is None
             or not any(t is not None for t in spec_tokens_all))
        and not any(t in _SPECIAL_SAMPLERS for t in sampler_types)
    )

    if not fast_path:
        # Slow path: build the dict via create_responses, msgpack-encode it.
        responses = batch.create_responses(sampling_results)
        results = [
            {
                "tokens": resp.tokens,
                "dists": resp.dists,
                "logits": getattr(resp, "logits", []) or [],
                "logprobs": getattr(resp, "logprobs", []) or [],
                "entropies": getattr(resp, "entropies", []) or [],
                "spec_tokens": getattr(resp, "spec_tokens", []) or [],
                "spec_positions": getattr(resp, "spec_positions", []) or [],
            }
            for resp in responses
        ]
        return write_response(
            {"results": results},
            dst_buf,
            msgpack_module,
        )

    # Fast path: counts come straight from `request_output_counts`, tokens
    # from `sampling_results['tokens']` (already in flat order).
    counts = batch.request_output_counts
    tokens = sampling_results["tokens"]

    if not isinstance(counts, np.ndarray) or counts.dtype != np.uint32:
        counts = np.asarray(counts, dtype=np.uint32)
    if not isinstance(tokens, np.ndarray) or tokens.dtype != np.uint32:
        tokens = np.asarray(tokens, dtype=np.uint32)

    n_req = counts.size
    n_tokens = tokens.size
    body_size = (n_req + n_tokens) * 4
    total = RESP_HEADER_SIZE + body_size
    dst_capacity = len(dst_buf)
    if total > dst_capacity:
        raise ValueError(
            f"shmem response (flat-v2) {total} > dst {dst_capacity}; "
            f"raise the SHMEM_RESP_BUF constant in worker.py / device.rs"
        )

    _struct_header_pack(
        dst_buf, 0, RESP_MAGIC, RESP_MODE_FLAT, n_req, n_tokens
    )
    counts_off = RESP_HEADER_SIZE
    tokens_off = counts_off + n_req * 4
    if n_req > 0:
        dst_buf[counts_off:tokens_off] = counts.tobytes()
    if n_tokens > 0:
        dst_buf[tokens_off:total] = tokens.tobytes()

    return total


def write_response(result: dict, dst_buf, msgpack_module=None) -> int:
    """Write a worker `result` dict directly into the shmem response buffer.

    `result` is the return value of `_handle_fire_batch`, shaped as
    `{"results": [{"tokens": [...], "dists": [...], ...}, ...]}`.
    `dst_buf` is a writable buffer-protocol object (e.g.
    `(c_uint8 * N).from_address(...)`).

    Returns: number of bytes written.
    """
    global _struct_header_pack
    if _struct_header_pack is None:
        import struct as _struct
        _struct_header_pack = _struct.Struct("<IIII").pack_into

    results = result["results"]
    n_req = len(results)

    # Single-pass advanced detection + counts + flat token list.
    # Cheaper than two passes (one to detect, one to count). The truthy
    # checks short-circuit on common-case empty advanced fields.
    counts = [0] * n_req
    flat_tokens: list = []
    flat_extend = flat_tokens.extend
    advanced = False
    for i, r in enumerate(results):
        if (r["dists"] or r["logits"] or r["logprobs"] or r["entropies"]
                or r["spec_tokens"] or r["spec_positions"]):
            advanced = True
            break
        t = r["tokens"]
        counts[i] = len(t)
        flat_extend(t)

    dst_capacity = len(dst_buf)

    if advanced:
        # Fallback path. Defer to msgpack of the full dict.
        if msgpack_module is None:
            import msgpack as msgpack_module
        body = msgpack_module.packb(result)
        body_len = len(body)
        total = RESP_HEADER_SIZE + body_len
        if total > dst_capacity:
            raise ValueError(
                f"shmem response (msgpack-fallback) {total} > dst {dst_capacity}; "
                f"raise the SHMEM_RESP_BUF constant in worker.py / device.rs"
            )
        _struct_header_pack(dst_buf, 0, RESP_MAGIC, RESP_MODE_MSGPACK, n_req, 0)
        dst_buf[RESP_HEADER_SIZE:total] = body
        return total

    # Token-only flat path.
    total_tokens = len(flat_tokens)
    body_size = (n_req + total_tokens) * 4
    total = RESP_HEADER_SIZE + body_size
    if total > dst_capacity:
        raise ValueError(
            f"shmem response (flat) {total} > dst {dst_capacity}; "
            f"raise the SHMEM_RESP_BUF constant in worker.py / device.rs"
        )

    _struct_header_pack(
        dst_buf, 0, RESP_MAGIC, RESP_MODE_FLAT, n_req, total_tokens
    )
    counts_off = RESP_HEADER_SIZE
    tokens_off = counts_off + n_req * 4

    # Single numpy array build + frombuffer assignment for the body. We
    # avoid creating a numpy view of dst_buf (which has noticeable overhead
    # per call); ctypes Arrays support slice-assignment from a bytes-like.
    if n_req > 0:
        counts_bytes = np.array(counts, dtype=np.uint32).tobytes()
        dst_buf[counts_off:tokens_off] = counts_bytes
    if total_tokens > 0:
        tokens_bytes = np.array(flat_tokens, dtype=np.uint32).tobytes()
        dst_buf[tokens_off:total] = tokens_bytes

    return total

File: runtime/src/shmem_schema.rs
//! Zero-copy flat schema for `BatchedForwardPassRequest` over shmem.
//!
//! ## Layout
//!
//! ```text
//! [512-byte header]
//!   0:  u32 magic = 0x42504951 ("BPIQ")
//!   4:  u32 schema_version = 2
//!   8:  u32 device_id
//!   12: u32 flags  (bit 0 = single_token_mode)
//!   16: u32 num_arrays = 29
//!   20: u32 reserved
//!   24: u64 reserved
//!   32: 29 × (u32 offset, u32 len_in_elements) = 232 bytes
//!   264..512: reserved (room for future array slots without resizing)
//!   512: array data (concatenated, each array 8-byte aligned)
//! ```
//!
//! Byte order: little-endian, matches host. We assume both processes are
//! on x86_64 LE.
//!
//! v2 added `A_PREDICT_FLAGS` (one u8 per request) for pass-level
//! speculative execution and bumped HEADER_SIZE 256 → 512 so the
//! offset/len table has room (32 + 29*8 = 264 > 256). See
//! SPECULATIVE_EXECUTION_DESIGN.md.

use crate::inference::request::BatchedForwardPassRequest;

pub const MAGIC: u32 = 0x42504951; // 'BPIQ'
pub const SCHEMA_VERSION: u32 = 2;
pub const HEADER_SIZE: usize = 512;
pub const NUM_ARRAYS: usize = 29;

const FIXED_HEADER: usize = 32;

// Array indices (must match Python side).
pub const A_TOKEN_IDS: usize = 0;
pub const A_POSITION_IDS: usize = 1;
pub const A_KV_PAGE_INDICES: usize = 2;
pub const A_KV_PAGE_INDPTR: usize = 3;
pub const A_KV_LAST_PAGE_LENS: usize = 4;
pub const A_QO_INDPTR: usize = 5;
pub const A_FLATTENED_MASKS: usize = 6;
pub const A_MASK_INDPTR: usize = 7;
pub const A_LOGIT_MASKS: usize = 8;
pub const A_LOGIT_MASK_INDPTR: usize = 9;
pub const A_SAMPLING_INDICES: usize = 10;
pub const A_SAMPLING_INDPTR: usize = 11;
pub const A_SAMPLER_TEMPERATURES: usize = 12; // f32
pub const A_SAMPLER_TOP_K: usize = 13;
pub const A_SAMPLER_TOP_P: usize = 14; // f32
pub const A_SAMPLER_MIN_P: usize = 15; // f32
pub const A_SAMPLER_TYPES: usize = 16;
pub const A_SAMPLER_SEEDS: usize = 17;
pub const A_REQUEST_NUM_SAMPLERS: usize = 18;
pub const A_SAMPLER_LABEL_IDS: usize = 19;
pub const A_SAMPLER_LABEL_INDPTR: usize = 20;
pub const A_ADAPTER_INDICES: usize = 21; // i64, -1 = None
pub const A_ADAPTER_SEEDS: usize = 22; // i64, i64::MIN = None
pub const A_SPEC_TOKEN_IDS: usize = 23;
pub const A_SPEC_POSITION_IDS: usize = 24;
pub const A_SPEC_INDPTR: usize = 25;
pub const A_OUTPUT_SPEC_FLAGS: usize = 26; // u8
pub const A_CONTEXT_IDS: usize = 27; // u64
pub const A_PREDICT_FLAGS: usize = 28; // u8 (v2)

const ELEM_SIZE: [usize; NUM_ARRAYS] = [
    4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
    4, 4, 4, 4, 4, 4, 4, 4, 4, 8, 8, 4, 4, 4, 1, 8, 1,
];

#[inline]
fn align_up(n: usize, a: usize) -> usize {
    (n + a - 1) & !(a - 1)
}

/// Serialize a `BatchedForwardPassRequest` into `buf`. Returns the number
/// of bytes written. Returns `Err` if the buffer is too small.
pub fn write_request(req: &BatchedForwardPassRequest, buf: &mut [u8]) -> anyhow::Result<usize> {
    if buf.len() < HEADER_SIZE {
        anyhow::bail!("shmem buffer {} < header size {}", buf.len(), HEADER_SIZE);
    }

    // Compute lengths per array.
    let lens: [usize; NUM_ARRAYS] = [
        req.token_ids.0.len(),
        req.position_ids.0.len(),
        req.kv_page_indices.0.len(),
        req.kv_page_indptr.0.len(),
        req.kv_last_page_lens.0.len(),
        req.qo_indptr.0.len(),
        req.flattened_masks.0.len(),
        req.mask_indptr.0.len(),
        req.logit_masks.0.len(),
        req.logit_mask_indptr.0.len(),
        req.sampling_indices.0.len(),
        req.sampling_indptr.0.len(),
        req.sampler_temperatures.0.len(),
        req.sampler_top_k.0.len(),
        req.sampler_top_p.0.len(),
        req.sampler_min_p.0.len(),
        req.sampler_types.0.len(),
        req.sampler_seeds.0.len(),
        req.request_num_samplers.0.len(),
        req.sampler_label_ids.0.len(),
        req.sampler_label_indptr.0.len(),
        req.adapter_indices.len(),
        req.adapter_seeds.len(),
        req.spec_token_ids.0.len(),
        req.spec_position_ids.0.len(),
        req.spec_indptr.0.len(),
        req.output_spec_flags.len(),
        req.context_ids.len(),
        req.predict_flags.len(),
    ];

    // Compute offsets.
    let mut offsets = [0u32; NUM_ARRAYS];
    let mut cursor = HEADER_SIZE;
    for i in 0..NUM_ARRAYS {
        cursor = align_up(cursor, 8);
        if cursor > u32::MAX as usize {
            anyhow::bail!("array offset overflow");
        }
        offsets[i] = cursor as u32;
        cursor += lens[i] * ELEM_SIZE[i];
    }
    let total = cursor;
    if total > buf.len() {
        anyhow::bail!("shmem buffer {} < required {}", buf.len(), total);
    }

    // Write header.
    let h = &mut buf[..HEADER_SIZE];
    h[0..4].copy_from_slice(&MAGIC.to_le_bytes());
    h[4..8].copy_from_slice(&SCHEMA_VERSION.to_le_bytes());
    h[8..12].copy_from_slice(&(req.device_id as u32).to_le_bytes());
    let flags: u32 = if req.single_token_mode { 1 } else { 0 };
    h[12..16].copy_from_slice(&flags.to_le_bytes());
    h[16..20].copy_from_slice(&(NUM_ARRAYS as u32).to_le_bytes());
    h[20..24].copy_from_slice(&0u32.to_le_bytes()); // reserved
    h[24..32].copy_from_slice(&0u64.to_le_bytes()); // reserved

    // Write the offset/len table.
    for i in 0..NUM_ARRAYS {
        let base = FIXED_HEADER + i * 8;
        h[base..base + 4].copy_from_slice(&offsets[i].to_le_bytes());
        h[base + 4..base + 8].copy_from_slice(&(lens[i] as u32).to_le_bytes());
    }

    // Write array bodies. Use bytemuck to splat slices.
    // Fixed mapping arms: each closure writes the appropriate type.
    macro_rules! write_u32 {
        ($idx:expr, $slice:expr) => {{
            let off = offsets[$idx] as usize;
            let nbytes = lens[$idx] * 4;
            let bytes: &[u8] = bytemuck::cast_slice($slice);
            buf[off..off + nbytes].copy_from_slice(bytes);
        }};
    }
    macro_rules! write_f32 {
        ($idx:expr, $slice:expr) => {{
            let off = offsets[$idx] as usize;
            let nbytes = lens[$idx] * 4;
            let bytes: &[u8] = bytemuck::cast_slice($slice);
            buf[off..off + nbytes].copy_from_slice(bytes);
        }};
    }

    write_u32!(A_TOKEN_IDS, &req.token_ids.0);
    write_u32!(A_POSITION_IDS, &req.position_ids.0);
    write_u32!(A_KV_PAGE_INDICES, &req.kv_page_indices.0);
    write_u32!(A_KV_PAGE_INDPTR, &req.kv_page_indptr.0);
    write_u32!(A_KV_LAST_PAGE_LENS, &req.kv_last_page_lens.0);
    write_u32!(A_QO_INDPTR, &req.qo_indptr.0);
    write_u32!(A_FLATTENED_MASKS, &req.flattened_masks.0);
    write_u32!(A_MASK_INDPTR, &req.mask_indptr.0);
    write_u32!(A_LOGIT_MASKS, &req.logit_masks.0);
    write_u32!(A_LOGIT_MASK_INDPTR, &req.logit_mask_indptr.0);
    write_u32!(A_SAMPLING_INDICES, &req.sampling_indices.0);
    write_u32!(A_SAMPLING_INDPTR, &req.sampling_indptr.0);
    write_f32!(A_SAMPLER_TEMPERATURES, &req.sampler_temperatures.0);
    write_u32!(A_SAMPLER_TOP_K, &req.sampler_top_k.0);
    write_f32!(A_SAMPLER_TOP_P, &req.sampler_top_p.0);
    write_f32!(A_SAMPLER_MIN_P, &req.sampler_min_p.0);
    write_u32!(A_SAMPLER_TYPES, &req.sampler_types.0);
    write_u32!(A_SAMPLER_SEEDS, &req.sampler_seeds.0);
    write_u32!(A_REQUEST_NUM_SAMPLERS, &req.request_num_samplers.0);
    write_u32!(A_SAMPLER_LABEL_IDS, &req.sampler_label_ids.0);
    write_u32!(A_SAMPLER_LABEL_INDPTR, &req.sampler_label_indptr.0);

    // Adapter indices (i64, -1 = None). AdapterId is u64; we serialize as
    // i64 with -1 sentinel — preserves the valid range used in practice.
    {
        let off = offsets[A_ADAPTER_INDICES] as usize;
        for (k, opt) in req.adapter_indices.iter().enumerate() {
            let v: i64 = match opt {
                Some(id) => *id as i64,
                None => -1,
            };
            buf[off + k * 8..off + (k + 1) * 8].copy_from_slice(&v.to_le_bytes());
        }
    }
    // Adapter seeds (i64, i64::MIN = None).
    {
        let off = offsets[A_ADAPTER_SEEDS] as usize;
        for (k, opt) in req.adapter_seeds.iter().enumerate() {
            let v: i64 = opt.unwrap_or(i64::MIN);
            buf[off + k * 8..off + (k + 1) * 8].copy_from_slice(&v.to_le_bytes());
        }
    }

    write_u32!(A_SPEC_TOKEN_IDS, &req.spec_token_ids.0);
    write_u32!(A_SPEC_POSITION_IDS, &req.spec_position_ids.0);
    write_u32!(A_SPEC_INDPTR, &req.spec_indptr.0);

    // output_spec_flags as u8.
    {
        let off = offsets[A_OUTPUT_SPEC_FLAGS] as usize;
        for (k, b) in req.output_spec_flags.iter().enumerate() {
            buf[off + k] = if *b { 1 } else { 0 };
        }
    }

    // context_ids as u64.
    {
        let off = offsets[A_CONTEXT_IDS] as usize;
        let bytes: &[u8] = bytemuck::cast_slice(&req.context_ids);
        let nbytes = req.context_ids.len() * 8;
        buf[off..off + nbytes].copy_from_slice(bytes);
    }

    // predict_flags as u8 (v2).
    {
        let off = offsets[A_PREDICT_FLAGS] as usize;
        for (k, b) in req.predict_flags.iter().enumerate() {
            buf[off + k] = if *b { 1 } else { 0 };
        }
    }

    Ok(total)
}


// =============================================================================
// Response schema
// =============================================================================

pub const RESP_MAGIC: u32 = 0x42504953; // 'BPIS'
pub const RESP_HEADER_SIZE: usize = 16;
pub const RESP_MODE_FLAT: u32 = 0;
pub const RESP_MODE_MSGPACK: u32 = 1;

use crate::inference::request::{BatchedForwardPassResponse, ForwardPassResponse};

/// Parse a `BatchedForwardPassResponse` from a shmem buffer.
///
/// Recognizes either the flat token-only schema (fast path) or a
/// msgpack-encoded fallback. Returns the parsed struct.
pub fn read_response(buf: &[u8]) -> anyhow::Result<BatchedForwardPassResponse> {
    if buf.len() < RESP_HEADER_SIZE {
        anyhow::bail!("response buffer too small ({} < {})", buf.len(), RESP_HEADER_SIZE);
    }
    let magic = u32::from_le_bytes(buf[0..4].try_into().unwrap());
    if magic != RESP_MAGIC {
        anyhow::bail!(
            "response magic mismatch: 0x{:08x} != 0x{:08x}",
            magic, RESP_MAGIC
        );
    }
    let mode = u32::from_le_bytes(buf[4..8].try_into().unwrap());
    let n_req = u32::from_le_bytes(buf[8..12].try_into().unwrap()) as usize;
    let total_tokens = u32::from_le_bytes(buf[12..16].try_into().unwrap()) as usize;

    if mode == RESP_MODE_MSGPACK {
        // Fallback path — body is msgpack-encoded.
        let body = &buf[RESP_HEADER_SIZE..];
        let resp: BatchedForwardPassResponse = rmp_serde::from_slice(body)
            .map_err(|e| anyhow::anyhow!("failed to deserialize msgpack response: {e}"))?;
        return Ok(resp);
    }
    if mode != RESP_MODE_FLAT {
        anyhow::bail!("unknown response mode {}", mode);
    }

    // Flat path: counts table, then tokens table.
    let counts_off = RESP_HEADER_SIZE;
    let tokens_off = counts_off + n_req * 4;
    let needed = tokens_off + total_tokens * 4;
    if buf.len() < needed {
        anyhow::bail!("response body truncated: have {}, need {}", buf.len(), needed);
    }

    let counts: &[u32] = bytemuck::cast_slice(&buf[counts_off..counts_off + n_req * 4]);
    let tokens_all: &[u32] = bytemuck::cast_slice(&buf[tokens_off..tokens_off + total_tokens * 4]);

    let mut results: Vec<ForwardPassResponse> = Vec::with_capacity(n_req);
    let mut cursor = 0usize;
    for i in 0..n_req {
        let n = counts[i] as usize;
        let toks = tokens_all[cursor..cursor + n].to_vec();
        cursor += n;
        results.push(ForwardPassResponse {
            tokens: toks,
            dists: Vec::new(),
            logits: Vec::new(),
            logprobs: Vec::new(),
            entropies: Vec::new(),
            spec_tokens: Vec::new(),
            spec_positions: Vec::new(),
        });
    }

    Ok(BatchedForwardPassResponse { results })
}
